### PR TITLE
[mle] support for long routes

### PR DESCRIPTION
**Pull request by Nopik**
This PR adds support for long routes (with routing cost > 16), as per discussion in #2758 

Current status: Ready for review / merge

### CODE REVIEW

**Comment by Nopik** [comment]
The implementation seem to be pretty much finished. Missing bits:
- tests (toranj doesnt work on any of my machines, I will need to figure it out first)
- support for R bit in Security Policy TLV (waiting for #2778)

So far it does seem to work on my basic local tests. I did noticed, however, that with 32 routers arranged in line, and having router upgrade & downgrade thresholds to 33, still some nodes occasionally are creating new partitions, then re-join. I think it might be some parent timeout, will investigate it as well.

Please review.

**Comment by Nopik** [comment]
I've realized, that my approach to modify `RouteTlv` directly is a bad choice. It will end up having one tlv of 2 different lengths and different semantics, and it will be hard to distinguish how to interpret incoming tlv. I will create new tlv, e.g. `ExtRouteTlv` with all the new implementation so `RouteTlv` will return to its pristine version instead of being littered with `#if`s. Then `FillRouteTlv` and `ProcessRouteTlv` will need to be adjusted slightly to handle the new cases.

The idea & handling already present in the code are still good for review, though.

**Comment by jwhui** [comment]
Overall, I think this looks good.  I agree that defining a new TLV is preferable.

Regarding the network partitioning that you observe, did you try modifying the MLE Advertisement period to something shorter (as discussed [here](https://groups.google.com/d/msg/openthread-users/F6YY_lQiTZg/QOUAPa6LAAAJ))?

File: configure.ac [hunk 1]
```diff
@@ -1266,6 +1266,36 @@ AC_SUBST(OPENTHREAD_ENABLE_SERVICE)
 AM_CONDITIONAL([OPENTHREAD_ENABLE_SERVICE], [test "${enable_service}" = "yes"])
 AC_DEFINE_UNQUOTED([OPENTHREAD_ENABLE_SERVICE],[${OPENTHREAD_ENABLE_SERVICE}],[Define to 1 if you want to enable Service])
 
+#
+# Long routes
+#
+
+AC_ARG_ENABLE(long_routes,
```
Comment (jwhui) [inline configure.ac]: I don't think the configure option here is necessary.   We could add a definition directly in [src/core/openthread-core-default-config.h](https://github.com/openthread/openthread/blob/master/src/core/openthread-core-default-config.h)

Comment (Nopik) [reply configure.ac]: Ok, I will remove it.

**Comment by Nopik** [comment]
No, I haven't modified MLE Advertisement period yet. That was my plan, though, for next steps (after new TLV will get introduced). I will probably get back to it in few days.

**Comment by Nopik** [comment]
@jwhui how you prefer the new TLV to be introduced? The easiest path (but least elegant) would be just have thing like:

```
#ifndef LONG_ROUTES_ENABLED
class RouteTlv { ... }
#else
class RouteTlv { ... }
#endif
```
So, rest of the code won't need updates.

Another approach would be to have

```
#ifdef LONG_ROUTES_ENABLED
class LongRouteTlv { ... }
#endif
```
and then modify each and every instance of `RouteTlv` in the code (e.g. in `FillTlv()`) with `#ifdef / #else / #endif` stuff, which I think is ugly.

Yet another approach is to introduce some base class and name it `RouteTlv` so plain old TLV can inherit from it, just as new TLV, so the only `#ifdef` stuff will be during creation of new TLV (in `FillTlv()` IIRC).

Or, maybe some other approach should be used?

**Comment by jwhui** [comment]
@Nopik, thanks for sharing your thoughts.

Given that this feature is not specified by Thread and is "experimental" for now, I'm inclined to agree that there should be no resource overhead for those not using this feature.

Of the methods you listed, are you OK to move forward with the first approach (i.e. redefining the `RouteTlv` class) for now?  As you noted, this minimizes overall code updates.

**Comment by Nopik** [comment]
Yeah, I will continue with your preferred version. Indeed at the current state it seems to be the best way to move on. If the change will prove to be useful, the version from Thread 2.0 will probably be different anyway and/or require further code changes.

I will need some time to continue, but hopefully manage to get it done by the end of the month.

**Comment by Nopik** [comment]
@jwhui I think I've introduced all changes which we've talked about.

From my perspective this PR is ready for merge.

Once #2778 will get merged, it will be possible to ignore the R bit in SecurityPolicy TLV when long routes are enabled.

**Comment by jwhui** [comment]
@Nopik, thanks for the update!

Did you resolve the partitioning issue you noted above?

**Comment by Nopik** [comment]
@jwhui yes, simply reducing the max advertise interval as you suggested did the trick.

**Pull request merged**